Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,730 @@
+  // dns request
+  StackRef* ref = GetAvaibleSlot(udpdata->source_port);
+  if (ref->ip == 0)
+  {
+    ref->ip = *(u32*)ipdata->destination_addr;  // change for ip
+    ref->local = udpdata->source_port;
+    ref->remote = udpdata->destination_port;
+    ref->type = 17;
+    ref->bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+    ref->my_mac = fake_mac;
+    ref->from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    ref->from.sin_port = udpdata->destination_port;
+    ref->to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    ref->to.sin_port = udpdata->source_port;
+    ref->udp_socket.setBlocking(false);
+    if (ref->udp_socket.bind(htons(udpdata->source_port)) != sf::Socket::Done &&
+        ref->udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+    {
+      ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+      return;
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 1900)
+  {
+    InitUDPPort(26512);                                 // MK DD and 1080
+    InitUDPPort(26502);                                 // Air Ride
+    if (*(u32*)ipdata->destination_addr == 0xFAFFFFEF)  // force real broadcast
+      *(u32*)ipdata->destination_addr = 0xFFFFFFFF;     // Multi cast cannot be read
+    if (udpdata->length > 150)
+    {
+      // Quick hack to unlock the connection, throw it back at him
+      Common::EthernetHeader* hwpart = (Common::EthernetHeader*)m_in_frame.get();
+      Common::IPv4Header* ippart = (Common::IPv4Header*)&m_in_frame[14];
+      std::memcpy(m_in_frame.get(), hwdata, htons(ipdata->total_len) + 14);
+      hwpart->destination = hwdata->source;
+      hwpart->source = hwdata->destination;
+      *(u32*)ippart->destination_addr = *(u32*)ipdata->source_addr;
+      *(u32*)ippart->source_addr = *(u32*)ipdata->destination_addr;
+      WriteToQueue(m_in_frame.get(), htons(ipdata->total_len) + 14);
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 53)
+  {
+    target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+  }
+  else
+  {
+    target = sf::IpAddress(Common::swap32(*(u32*)ipdata->destination_addr));
+  }
+  ref->udp_socket.send(data, Common::swap16(udpdata->length) - 8, target,
+                       Common::swap16(udpdata->destination_port));
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  Common::EthernetHeader* hwdata;
+  Common::IPv4Header* ipdata;
+  Common::UDPHeader* udpdata;
+  Common::TCPHeader* tcpdata;
+  int offset = 0;
+
+  std::memcpy(m_out_frame.get(), frame, size);
+
+  std::lock_guard<std::mutex> lock(mtx);
+
+  // handle the packet data
+  hwdata = (Common::EthernetHeader*)m_out_frame.get();
+  if (hwdata->ethertype == 0x08)  // IPV4
+  {
+    // IP sub
+    ipdata = (Common::IPv4Header*)&m_out_frame[14];
+    switch (ipdata->protocol)
+    {
+    case IPPROTO_UDP:
+      offset = 14 + (ipdata->version_ihl & 0xf) * 4;
+      udpdata = (Common::UDPHeader*)&m_out_frame[offset];
```

Comment:
```C++
Common::UDPHeader* udpdata = (Common::UDPHeader*)&m_out_frame[offset];
```

Reply:
tryed it, VS dont like it, you cant make declaration in a case so i left the declaration beafor the switch

Reply:
You have to brace the case statement, ie:

~~~
switch (a) {
 case 1: {
  int local_variable = 2;
 }
}
~~~

Reply:
Also this looks quite sketchy, not sure if it's actually UB but you probably should memcpy to a local UDPHeader or something instead.

Reply:
thx for the info. About doing a local copy, i dont see the need as a copied Header or in data pointer will give the same value, and it raise the cpu cost since u have to copy first from the same pointer source.